cirq/protocols/apply_mixture_protocol_test.py

```diff
@@ -0,0 +1,330 @@
+
+                cirq.apply_matrix_to_slices(args.auxiliary_buffer1,
+                                            np.conjugate(krauss_op),
+                                            [zero_right, one_right],
+                                            out=args.target_tensor)
+                args.out_buffer += args.target_tensor
+            return args.out_buffer
+
+    rho = np.copy(x)
+    result = apply_mixture(HasApplyMixture(),
+                           rho, [0], [1],
+                           assert_result_is_out_buf=True)
+    np.testing.assert_almost_equal(result, x)
+
+
+def test_apply_mixture_inline():
+    x = np.array([[0, 1], [1, 0]], dtype=np.complex128)
+
+    class HasApplyMixture():
+
+        def _apply_mixture_(self, args: cirq.ApplyMixtureArgs):
+            args.target_tensor = 0.5 * args.target_tensor + 0.5 * np.dot(
+                np.dot(x, args.target_tensor), x)
+            return args.target_tensor
+
+    rho = np.copy(x)
+    result = apply_mixture(HasApplyMixture(), rho, [0], [1])
+    np.testing.assert_almost_equal(result, x)
+
+
+def test_apply_mixture_returns_aux_buffer():
+    rho = np.array([[1, 0], [0, 0]], dtype=np.complex128)
+
+    class ReturnsAuxBuffer0():
+
+        def _apply_mixture_(self, args: cirq.ApplyMixtureArgs):
+            return args.auxiliary_buffer0
+
+    with pytest.raises(AssertionError, match='ReturnsAuxBuffer0'):
+        _ = apply_mixture(ReturnsAuxBuffer0(), rho, [0], [1])
+
+    class ReturnsAuxBuffer1():
+
+        def _apply_mixture_(self, args: cirq.ApplyMixtureArgs):
+            return args.auxiliary_buffer1
+
+    with pytest.raises(AssertionError, match='ReturnsAuxBuffer1'):
+        _ = apply_mixture(ReturnsAuxBuffer1(), rho, [0], [1])
+
+
+def test_apply_mixture_simple_wavefunction():
+    for _ in range(25):
+        state = cirq.testing.random_superposition(2)
+        u1 = cirq.testing.random_unitary(2)
+        u2 = np.array([[0, 1], [1, 0]], dtype=np.complex128)
+        p1 = np.random.random()
+        p2 = 1 - p1
+
+        expected = p1 * np.dot(u1, state) + p2 * np.dot(u2, state)
+
+        class HasMixture():
+
+            def _mixture_(self):
+                return ((p1, u1), (p2, cirq.X))
+
+        result = apply_mixture(HasMixture(),
+                               state, [0],
+                               None,
+                               assert_result_is_out_buf=True)
+
+        np.testing.assert_almost_equal(result, expected)
+
+
+def test_apply_mixture_simple_split_fallback():
+    x = np.array([[0, 1], [1, 0]], dtype=np.complex128)
+
+    class HasMixture():
+
+        def _mixture_(self):
+            return ((0.5, np.eye(2, dtype=np.complex128)), (0.5, cirq.X))
+
+    rho = np.copy(x)
+    result = apply_mixture(HasMixture(),
+                           rho, [0], [1],
+                           assert_result_is_out_buf=True)
+    np.testing.assert_almost_equal(result, x)
```

Comment:
Every call to `apply_mixture` is followed by an assertion about what the output is equal to. Fold the output almost equal check into the method and rename it to `assert_apply_mixture_returns`.

Reply:
changed